vcs/git (native): No support for annotated tags.

If you look at `GetCommit` and `FileSystem` implementations of `vcs/git` package, they both use `LookupCommit`, which doesn't work when the object is an annotated tag rather than a commit.

`ResolveTag` returns the tag's `CommitId`, rather than its parent. So I see no way to get a working `FileSystem` for a given annotated tag.

The test program from sourcegraph/vcsstore#11 updated to use the new import paths and to try to walk the filesystem (source [here](http://play.golang.org/p/4n_PalnY5K)), now prints:

```Go
commit = (*vcs.Commit)(&vcs.Commit{
	ID: (vcs.CommitID)("f8d048baeca3571b825c647ce6bdc59f9fbf004f"),
	Author: (vcs.Signature)(vcs.Signature{
		Name:  (string)(""),
		Email: (string)(""),
		Date: (pbtypes.Timestamp)(pbtypes.Timestamp{
			Seconds: (int64)(0),
			Nanos:   (int32)(0),
		}),
	}),
	Committer: (*vcs.Signature)(nil),
	Message:   (string)("Version 1 Release"),
	Parents:   ([]vcs.CommitID)(nil),
})
2016/02/03 00:16:02 can't stat file /: commit.Author was nil, which may be unexpected
```

Which seems reasonable and better than the previous output:

```
panic: The requested type does not match the type in the ODB
```

But it may not be enough to resolve the underlying issue, it's still not possible to access the `FileSystem` of an annotated tag, is there?

--- a/vcs/git/repo.go
+++ b/vcs/git/repo.go
@@ -102,6 +102,10 @@ func (r *Repository) ResolveRevision(spec string) (vcs.CommitID, error) {
 // ResolveTag returns the tag with the given name, or
 // ErrTagNotFound if no such tag exists.
 func (r *Repository) ResolveTag(name string) (vcs.CommitID, error) {
+	// TODO: Implement non-fallback that dereferences annotated tags
+	//       consistently with gitcmd version. See issue #24.
+	return r.Repository.ResolveTag(name)
+
 	id, err := r.repo.GetCommitIdOfTag(name)
 	if _, ok := err.(git.RefNotFound); ok {
 		return "", vcs.ErrTagNotFound

--- a/vcs/gitcmd/repo.go
+++ b/vcs/gitcmd/repo.go
@@ -14,7 +14,6 @@ import (
 	"path/filepath"
 	"regexp"
 	"runtime"
-	"sort"
 	"strconv"
 	"strings"
 	"sync"
@@ -326,24 +325,33 @@ func (r *Repository) Tags() ([]*vcs.Tag, error) {
 		return nil, err
 	}
 
-	tags := make([]*vcs.Tag, len(refs))
-	for i, ref := range refs {
-		tags[i] = &vcs.Tag{
+	var tags []*vcs.Tag
+	for _, ref := range refs {
+		if strings.HasSuffix(ref[1], "^{}") {
+			// A dereferenced tag always follows the non-dereferenced one,
+			// and is shown with "^{}" appended. E.g.:
+			//
+			// 	3521017556c5de4159da4615a39fa4d5d2c279b5 refs/tags/v0.99.9c
+			// 	6ddc0964034342519a87fe013781abf31c6db6ad refs/tags/v0.99.9c^{}
+			// 	055e4ae3ae6eb344cbabf2a5256a49ea66040131 refs/tags/v1.0rc4
+			// 	423325a2d24638ddcc82ce47be5e40be550f4507 refs/tags/v1.0rc4^{}
+			//
+			// So override CommitID of previous tag to its dereferenced value.
+			tags[len(tags)-1].CommitID = vcs.CommitID(ref[0])
+			continue
+		}
+		tags = append(tags, &vcs.Tag{
 			Name:     strings.TrimPrefix(ref[1], "refs/tags/"),
 			CommitID: vcs.CommitID(ref[0]),
-		}
+		})
 	}
 	return tags, nil
 }
 
-type byteSlices [][]byte
-
-func (p byteSlices) Len() int           { return len(p) }
-func (p byteSlices) Less(i, j int) bool { return bytes.Compare(p[i], p[j]) < 0 }
-func (p byteSlices) Swap(i, j int)      { p[i], p[j] = p[j], p[i] }
-
-func (r *Repository) showRef(arg string) ([][2]string, error) {
-	cmd := exec.Command("git", "show-ref", arg)
+// showRef calls "git show-ref {filter} --dereference" and splits
+// the output by line. filter can be one of "--heads" or "--tags".
+func (r *Repository) showRef(filter string) ([][2]string, error) {
+	cmd := exec.Command("git", "show-ref", filter, "--dereference")
 	cmd.Dir = r.Dir
 	out, err := cmd.CombinedOutput()
 	if err != nil {
@@ -352,12 +360,11 @@ func (r *Repository) showRef(arg string) ([][2]string, error) {
 		if exitStatus(err) == 1 && len(out) == 0 {
 			return nil, nil
 		}
-		return nil, fmt.Errorf("exec `git show-ref %s` in %s failed: %s. Output was:\n\n%s", arg, r.Dir, err, out)
+		return nil, fmt.Errorf("exec `git show-ref %s --dereference` in %s failed: %s. Output was:\n\n%s", filter, r.Dir, err, out)
 	}
 
 	out = bytes.TrimSuffix(out, []byte("\n")) // remove trailing newline
 	lines := bytes.Split(out, []byte("\n"))
-	sort.Sort(byteSlices(lines)) // sort for consistency
 	refs := make([][2]string, len(lines))
 	for i, line := range lines {
 		if len(line) <= 41 {

--- a/vcs/repository_test.go
+++ b/vcs/repository_test.go
@@ -151,6 +151,7 @@ func TestRepository_ResolveRevision(t *testing.T) {
 
 	gitCommands := []string{
 		"GIT_COMMITTER_NAME=a GIT_COMMITTER_EMAIL=a@a.com GIT_COMMITTER_DATE=2006-01-02T15:04:05Z git commit --allow-empty -m foo --author='a <a@a.com>' --date 2006-01-02T15:04:05Z",
+		"GIT_COMMITTER_NAME=a GIT_COMMITTER_EMAIL=a@a.com GIT_COMMITTER_DATE=2006-01-02T15:04:06Z git tag -a ta -m bar",
 	}
 	hgCommands := []string{
 		"touch --date=2006-01-02T15:04:05Z f || touch -t " + times[0] + " f",
@@ -169,6 +170,16 @@ func TestRepository_ResolveRevision(t *testing.T) {
 			spec:         "master",
 			wantCommitID: "ea167fe3d76b1e5fd3ed8ca44cbd2fe3897684f8",
 		},
+		"git cmd complex": {
+			repo:         makeGitRepositoryCmd(t, gitCommands...),
+			spec:         "master@{0}",
+			wantCommitID: "ea167fe3d76b1e5fd3ed8ca44cbd2fe3897684f8",
+		},
+		"git cmd annotated tag": {
+			repo:         makeGitRepositoryCmd(t, gitCommands...),
+			spec:         "ta",
+			wantCommitID: "ea167fe3d76b1e5fd3ed8ca44cbd2fe3897684f8",
+		},
 		"git go-git": {
 			repo:         makeGitRepositoryGoGit(t, gitCommands...),
 			spec:         "master",
@@ -178,6 +189,11 @@ func TestRepository_ResolveRevision(t *testing.T) {
 			repo:         makeGitRepositoryGoGit(t, gitCommands...),
 			spec:         "master@{0}",
 			wantCommitID: "ea167fe3d76b1e5fd3ed8ca44cbd2fe3897684f8",
+		},
+		"git go-git annotated tag": {
+			repo:         makeGitRepositoryGoGit(t, gitCommands...),
+			spec:         "ta",
+			wantCommitID: "d07bdca22f4731f877b3b98177adc033c7a22263",
 		},
 		"hg native": {
 			repo:         makeHgRepositoryNative(t, hgCommands...),
@@ -294,6 +310,7 @@ func TestRepository_ResolveTag(t *testing.T) {
 	gitCommands := []string{
 		"GIT_COMMITTER_NAME=a GIT_COMMITTER_EMAIL=a@a.com GIT_COMMITTER_DATE=2006-01-02T15:04:05Z git commit --allow-empty -m foo --author='a <a@a.com>' --date 2006-01-02T15:04:05Z",
 		"git tag t",
+		"GIT_COMMITTER_NAME=a GIT_COMMITTER_EMAIL=a@a.com GIT_COMMITTER_DATE=2006-01-02T15:04:06Z git tag -a ta -m bar",
 	}
 	hgCommands := []string{
 		"touch --date=2006-01-02T15:04:05Z f || touch -t " + times[0] + " f",
@@ -308,15 +325,25 @@ func TestRepository_ResolveTag(t *testing.T) {
 		tag          string
 		wantCommitID vcs.CommitID
 	}{
-		"git cmd": {
+		"git cmd lightweight tag": {
 			repo:         makeGitRepositoryCmd(t, gitCommands...),
 			tag:          "t",
 			wantCommitID: "ea167fe3d76b1e5fd3ed8ca44cbd2fe3897684f8",
 		},
-		"git go-git": {
+		"git cmd annotated tag": {
+			repo:         makeGitRepositoryCmd(t, gitCommands...),
+			tag:          "ta",
+			wantCommitID: "ea167fe3d76b1e5fd3ed8ca44cbd2fe3897684f8",
+		},
+		"git go-git lightweight tag": {
 			repo:         makeGitRepositoryGoGit(t, gitCommands...),
 			tag:          "t",
 			wantCommitID: "ea167fe3d76b1e5fd3ed8ca44cbd2fe3897684f8",
+		},
+		"git go-git annotated tag": {
+			repo:         makeGitRepositoryGoGit(t, gitCommands...),
+			tag:          "ta",
+			wantCommitID: "d07bdca22f4731f877b3b98177adc033c7a22263",
 		},
 		"hg native": {
 			repo:         makeHgRepositoryNative(t, hgCommands...),
@@ -694,6 +721,7 @@ func TestRepository_Tags(t *testing.T) {
 		"GIT_COMMITTER_NAME=a GIT_COMMITTER_EMAIL=a@a.com GIT_COMMITTER_DATE=2006-01-02T15:04:05Z git commit --allow-empty -m foo --author='a <a@a.com>' --date 2006-01-02T15:04:05Z",
 		"git tag t0",
 		"git tag t1",
+		"GIT_COMMITTER_NAME=a GIT_COMMITTER_EMAIL=a@a.com GIT_COMMITTER_DATE=2006-01-02T15:04:06Z git tag -a ta -m bar",
 	}
 	hgCommands := []string{
 		"touch --date=2006-01-02T15:04:05Z f || touch -t " + times[0] + " f",
@@ -709,12 +737,20 @@ func TestRepository_Tags(t *testing.T) {
 		wantTags []*vcs.Tag
 	}{
 		"git cmd": {
-			repo:     makeGitRepositoryCmd(t, gitCommands...),
-			wantTags: []*vcs.Tag{{Name: "t0", CommitID: "ea167fe3d76b1e5fd3ed8ca44cbd2fe3897684f8"}, {Name: "t1", CommitID: "ea167fe3d76b1e5fd3ed8ca44cbd2fe3897684f8"}},
+			repo: makeGitRepositoryCmd(t, gitCommands...),
+			wantTags: []*vcs.Tag{
+				{Name: "t0", CommitID: "ea167fe3d76b1e5fd3ed8ca44cbd2fe3897684f8"},
+				{Name: "t1", CommitID: "ea167fe3d76b1e5fd3ed8ca44cbd2fe3897684f8"},
+				{Name: "ta", CommitID: "d07bdca22f4731f877b3b98177adc033c7a22263"},
+			},
 		},
 		"git go-git": {
-			repo:     makeGitRepositoryGoGit(t, gitCommands...),
-			wantTags: []*vcs.Tag{{Name: "t0", CommitID: "ea167fe3d76b1e5fd3ed8ca44cbd2fe3897684f8"}, {Name: "t1", CommitID: "ea167fe3d76b1e5fd3ed8ca44cbd2fe3897684f8"}},
+			repo: makeGitRepositoryGoGit(t, gitCommands...),
+			wantTags: []*vcs.Tag{
+				{Name: "t0", CommitID: "ea167fe3d76b1e5fd3ed8ca44cbd2fe3897684f8"},
+				{Name: "t1", CommitID: "ea167fe3d76b1e5fd3ed8ca44cbd2fe3897684f8"},
+				{Name: "ta", CommitID: "d07bdca22f4731f877b3b98177adc033c7a22263"},
+			},
 		},
 		"hg native": {
 			repo:     makeHgRepositoryNative(t, hgCommands...),
@@ -1540,6 +1576,111 @@ func TestRepository_FileSystem(t *testing.T) {
 		}
 		if file1Info := dir1Entries[0]; file1Info.Name() != "file1" {
 			t.Errorf("%s: got dir1 entry name == %q, want 'file1'", label, file1Info.Name())
+		}
+	}
+}
+
+// Test that it's possible to get at the underlying filesystem
+// of a git tag, both lightweight and annotated.
+// See issue https://github.com/sourcegraph/go-vcs/issues/24.
+func TestRepository_FileSystem_fromTag(t *testing.T) {
+	t.Parallel()
+
+	file1MTime, err := time.Parse(time.RFC3339, "2006-01-02T15:04:05Z")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	gitCommands := []string{
+		"echo -n infile1 > file1",
+		"touch --date=2006-01-02T15:04:05Z file1 || touch -t " + times[0] + " file1",
+		"git add file1",
+		"GIT_COMMITTER_NAME=a GIT_COMMITTER_EMAIL=a@a.com GIT_COMMITTER_DATE=2006-01-02T15:04:05Z git commit -m foo --author='a <a@a.com>' --date 2006-01-02T15:04:05Z",
+		"git tag t",
+		"GIT_COMMITTER_NAME=a GIT_COMMITTER_EMAIL=a@a.com GIT_COMMITTER_DATE=2006-01-02T15:04:06Z git tag -a ta -m bar",
+	}
+	tests := map[string]struct {
+		repo interface {
+			ResolveRevision(string) (vcs.CommitID, error)
+			FileSystem(vcs.CommitID) (vfs.FileSystem, error)
+		}
+		spec         string
+		wantCommitID vcs.CommitID
+	}{
+		"git cmd lightweight tag": {
+			repo:         makeGitRepositoryCmd(t, gitCommands...),
+			spec:         "t",
+			wantCommitID: "00143c1744a30ce602c4585b9c351c3c108841e3",
+		},
+		"git cmd annotated tag": {
+			repo:         makeGitRepositoryCmd(t, gitCommands...),
+			spec:         "ta",
+			wantCommitID: "", // TODO.
+		},
+		"git go-git lightweight tag": {
+			repo:         makeGitRepositoryGoGit(t, gitCommands...),
+			spec:         "t",
+			wantCommitID: "00143c1744a30ce602c4585b9c351c3c108841e3",
+		},
+		"git go-git annotated tag": {
+			repo:         makeGitRepositoryGoGit(t, gitCommands...),
+			spec:         "ta",
+			wantCommitID: "", // TODO.
+		},
+	}
+
+	for label, test := range tests {
+		if strings.Contains(label, "annotated") {
+			// TODO: Resolve issue 24, make the annotated test cases pass, and remove this skip.
+			t.Skip("skipping annotated tag test case, because it would fail; see issue https://github.com/sourcegraph/go-vcs/issues/24")
+		}
+
+		commitID, err := test.repo.ResolveRevision(test.spec)
+		if err != nil {
+			t.Errorf("%s: ResolveRevision: %s", label, err)
+			continue
+		}
+
+		if commitID != test.wantCommitID {
+			t.Errorf("%s: got commitID == %v, want %v", label, commitID, test.wantCommitID)
+		}
+
+		fs, err := test.repo.FileSystem(commitID)
+		if err != nil {
+			t.Errorf("%s: FileSystem: %s", label, err)
+			continue
+		}
+
+		// file1 should exist, contain "infile1", have the right mtime, and be a file.
+		file1, err := fs.Open("file1")
+		if err != nil {
+			t.Errorf("%s: fs.Open(file1): %s", label, err)
+			continue
+		}
+		file1Data, err := ioutil.ReadAll(file1)
+		if err != nil {
+			t.Errorf("%s: ReadAll(file1): %s", label, err)
+			continue
+		}
+		if !bytes.Equal(file1Data, []byte("infile1")) {
+			t.Errorf("%s: got file1Data == %q, want %q", label, string(file1Data), "infile1")
+		}
+		file1Info, err := fs.Stat("file1")
+		if err != nil {
+			t.Errorf("%s: fs.Stat(file1): %s", label, err)
+			continue
+		}
+		if !file1Info.Mode().IsRegular() {
+			t.Errorf("%s: file1 stat !IsRegular", label)
+		}
+		if name := file1Info.Name(); name != "file1" {
+			t.Errorf("%s: got file1 name %q, want 'file1'", label, name)
+		}
+		if size, want := file1Info.Size(), int64(len("infile1")); size != want {
+			t.Errorf("%s: got file1 size %d, want %d", label, size, want)
+		}
+		if mtime, want := file1Info.ModTime(), file1MTime; !mtime.Equal(want) {
+			t.Errorf("%s: got file1 mtime %v, want %v", label, mtime, want)
 		}
 	}
 }

--- a/vcs/repository_test.go
+++ b/vcs/repository_test.go
@@ -193,7 +193,7 @@ func TestRepository_ResolveRevision(t *testing.T) {
 		"git go-git annotated tag": {
 			repo:         makeGitRepositoryGoGit(t, gitCommands...),
 			spec:         "ta",
-			wantCommitID: "d07bdca22f4731f877b3b98177adc033c7a22263",
+			wantCommitID: "ea167fe3d76b1e5fd3ed8ca44cbd2fe3897684f8",
 		},
 		"hg native": {
 			repo:         makeHgRepositoryNative(t, hgCommands...),
@@ -343,7 +343,7 @@ func TestRepository_ResolveTag(t *testing.T) {
 		"git go-git annotated tag": {
 			repo:         makeGitRepositoryGoGit(t, gitCommands...),
 			tag:          "ta",
-			wantCommitID: "d07bdca22f4731f877b3b98177adc033c7a22263",
+			wantCommitID: "ea167fe3d76b1e5fd3ed8ca44cbd2fe3897684f8",
 		},
 		"hg native": {
 			repo:         makeHgRepositoryNative(t, hgCommands...),
@@ -741,7 +741,7 @@ func TestRepository_Tags(t *testing.T) {
 			wantTags: []*vcs.Tag{
 				{Name: "t0", CommitID: "ea167fe3d76b1e5fd3ed8ca44cbd2fe3897684f8"},
 				{Name: "t1", CommitID: "ea167fe3d76b1e5fd3ed8ca44cbd2fe3897684f8"},
-				{Name: "ta", CommitID: "d07bdca22f4731f877b3b98177adc033c7a22263"},
+				{Name: "ta", CommitID: "ea167fe3d76b1e5fd3ed8ca44cbd2fe3897684f8"},
 			},
 		},
 		"git go-git": {
@@ -749,7 +749,7 @@ func TestRepository_Tags(t *testing.T) {
 			wantTags: []*vcs.Tag{
 				{Name: "t0", CommitID: "ea167fe3d76b1e5fd3ed8ca44cbd2fe3897684f8"},
 				{Name: "t1", CommitID: "ea167fe3d76b1e5fd3ed8ca44cbd2fe3897684f8"},
-				{Name: "ta", CommitID: "d07bdca22f4731f877b3b98177adc033c7a22263"},
+				{Name: "ta", CommitID: "ea167fe3d76b1e5fd3ed8ca44cbd2fe3897684f8"},
 			},
 		},
 		"hg native": {
@@ -1615,7 +1615,7 @@ func TestRepository_FileSystem_fromTag(t *testing.T) {
 		"git cmd annotated tag": {
 			repo:         makeGitRepositoryCmd(t, gitCommands...),
 			spec:         "ta",
-			wantCommitID: "", // TODO.
+			wantCommitID: "00143c1744a30ce602c4585b9c351c3c108841e3",
 		},
 		"git go-git lightweight tag": {
 			repo:         makeGitRepositoryGoGit(t, gitCommands...),
@@ -1625,16 +1625,11 @@ func TestRepository_FileSystem_fromTag(t *testing.T) {
 		"git go-git annotated tag": {
 			repo:         makeGitRepositoryGoGit(t, gitCommands...),
 			spec:         "ta",
-			wantCommitID: "", // TODO.
+			wantCommitID: "00143c1744a30ce602c4585b9c351c3c108841e3",
 		},
 	}
 
 	for label, test := range tests {
-		if strings.Contains(label, "annotated") {
-			// TODO: Resolve issue 24, make the annotated test cases pass, and remove this skip.
-			t.Skip("skipping annotated tag test case, because it would fail; see issue https://github.com/sourcegraph/go-vcs/issues/24")
-		}
-
 		commitID, err := test.repo.ResolveRevision(test.spec)
 		if err != nil {
 			t.Errorf("%s: ResolveRevision: %s", label, err)

